src/jacobians.jl

```diff
@@ -0,0 +1,181 @@
+# Compute the Jacobian matrix of a real-valued callable f.
+function finite_difference_jacobian(f, x::AbstractArray{<:Number},
+    fdtype::DataType=Val{:central}, funtype::DataType=Val{:Real}, wrappertype::DataType=Val{:Default},
+    fx::Union{Void,AbstractArray{<:Number}}=nothing, epsilon::Union{Void,AbstractArray{<:Real}}=nothing, returntype=eltype(x),
+    df::Union{Void,AbstractArray{<:Number}}=nothing)
+
+    J = zeros(returntype, length(x), length(x))
+    finite_difference_jacobian!(J, f, x, fdtype, funtype, wrappertype, fx, epsilon, returntype, df)
+end
+
+function finite_difference_jacobian!(J::AbstractMatrix{<:Number}, df::AbstractVector, f, x::AbstractArray{<:Number},
+    fdtype::DataType=Val{:central}, funtype::DataType=Val{:Real}, wrappertype::DataType=Val{:Default},
+    fx::Union{Void,AbstractArray{<:Number}}=nothing, epsilon::Union{Void,AbstractArray{<:Number}}=nothing, returntype=eltype(x))
+
+    finite_difference_jacobian!(J, f, x, fdtype, funtype, wrappertype, fx, epsilon, returntype, df)
+end
+
+function finite_difference_jacobian!(J::AbstractMatrix{<:Number}, f, x::AbstractArray{<:Number},
+    fdtype::DataType=Val{:central}, funtype::DataType=Val{:Real}, wrappertype::DataType=Val{:Default},
+    fx::Union{Void,AbstractArray{<:Number}}=nothing, epsilon::Union{Void,AbstractArray{<:Number}}=nothing, returntype=eltype(x),
+    df::Union{Void,AbstractArray{<:Number}}=nothing)
+
+    finite_difference_jacobian!(J, f, x, fdtype, funtype, wrappertype, fx, epsilon, returntype, df)
+end
+
+function finite_difference_jacobian!(J::AbstractMatrix{<:Real}, f, x::AbstractArray{<:Real},
+    fdtype::DataType, ::Type{Val{:Real}}, ::Type{Val{:Default}},
+    fx::Union{Void,AbstractArray{<:Real}}=nothing, epsilon::Union{Void,AbstractArray{<:Real}}=nothing, returntype=eltype(x),
+    df::Union{Void,AbstractArray{<:Real}}=nothing)
+
+    # TODO: test and rework this
```

Comment:
test and rework what? Just this abstract dispatch?

Reply:
No, the implementation. It currently uses stuff like `J[:,i]`, which will fail for matrix types that don't allow indexing. It's not a regression because we've never supported it, but it's not quite general enough for completely arbitrary matrices yet, and hasn't been tested on sufficiently exotic types.

Reply:
Got it, thanks.